src/deepsparse/transformers/utils/decoder_kv_cache.py

```diff
@@ -0,0 +1,218 @@
+            that corresponds to the first token will be frozen.
+            This assures that, once the KV cache is full (there are no
+            "blank" entries), and we are removing the "oldest" entry
+            from the cache, we will nevertheless keep the cache entry
+            that corresponds to the BOS token in the sequence.
+            By default, is set to False.
+        """
+        self.session_id = session_id
+        self._state = state
+        # TODO: I am aware that I am altering the specs with this
+        # variable, but it makes more sense to set this as a boolean
+        # rather than a list of integers.
+        self._freeze_first_position = freeze_first_position
+        if self._kv_cache is not None:
+            self._kv_cache.reset(
+                num_tokens=num_tokens, freeze_first_position=self._freeze_first_position
+            )
+
+    def update_session(
+        self,
+        state: Dict[str, Any],
+        num_tokens: int,
+        input_ids_len: int,
+        ignore_generated: bool = False,
+    ):
+        """
+        Updating the session is identical with taking the kv cache
+        output of from the forward pass and restructuring it, so it
+        can be directly used as input for the next forward pass.
+
+        :param state: The state of the cache. This is a dictionary
+            that maps the name of the cache array to the cache array.
+            The cache tensor is a numpy array of shape
+            [batch_size, num_heads, sequence_length, hidden_size]
+        :param num_tokens: The number of tokens processed so far,
+            corresponding to the number of "non-blank" entries in the
+            kv cache array. Even though all numpy arrays in the state
+            of the cache have constant sequence length, this length
+            corresponds to the total "capacity" of the
+            cache. A portion of this memory can be "blank".
+        :param input_ids_len: The number of input ids in the current
+            input batch: (batch_size, length).
+            Corresponds to `input_ids.shape[1]`
+        :param ignore_generated: If set to True, the last entry in the
+            kv cache will be ignored. This is because this entry is
+            encoding the information about the newly generated logits,
+            which is not needed in the scenario, where we just prefill
+            the kv cache - populate the kv cache, without the immediate
+            goal of generating any new token.
+        """
+        total_cache_capacity = state[list(state.keys())[0]].shape[self._sequence_axis]
+        # total_capacity = num_tokens (num of non-blank tokens) +
+        # + num_padded_entries (num of blank tokens)
+        num_padded_entries = max(0, total_cache_capacity - num_tokens)
+        # we want to remove input_ids_len entries from the cache
+        # because len_input_ids + inp_cache_len = out_cache_len
+        num_entries_to_delete = input_ids_len
```

Comment:
sequence_len: int

output_kv_cache = (..., ..., sequence_len, ...)

num_tokens (all the tokens seen) is redundant given the input_ids ( given: sequence_len)